groovy-language-server/src/main/java/com/palantir/ls/server/groovy/GroovyTreeParser.java

```diff
@@ -150,66 +169,111 @@ public void parseAllSymbols() {
-        return typeReferences;
+    public Map<Location, Set<Location>> getReferences() {
+        return indexer.getReferences();
     }
 
     @Override
-    public Set<SymbolInformation> findReferences(ReferenceParams params) {
-        Set<SymbolInformation> symbols =
-                fileSymbols.get(Uris.resolveToRoot(workspaceRoot, params.getTextDocument().getUri()));
+    public Set<Location> findReferences(ReferenceParams params) {
+        URI paramsUri = Uris.resolveToRoot(workspaceRoot, params.getTextDocument().getUri());
+        Set<SymbolInformation> symbols = indexer.getFileSymbols().get(paramsUri);
         if (symbols == null) {
             return Sets.newHashSet();
         }
-        List<SymbolInformation> foundSymbolInformations = symbols.stream()
-                .filter(s -> Ranges.isValid(s.getLocation().getRange())
-                        && Ranges.contains(s.getLocation().getRange(), params.getPosition())
-                        && (s.getKind() == SymbolKind.Class || s.getKind() == SymbolKind.Interface
-                                || s.getKind() == SymbolKind.Enum))
-                // If there is more than one result, we want the symbol whose range starts the latest.
-                .sorted((s1, s2) -> Ranges.POSITION_COMPARATOR.reversed()
-                        .compare(s1.getLocation().getRange().getStart(), s2.getLocation().getRange().getStart()))
+        List<ReferenceLocation> foundReferencedLocations = symbols.stream()
+                .map(symbol -> symbol.getLocation())
+                .filter(l -> Ranges.isValid(l.getRange())
+                        && Ranges.contains(l.getRange(), params.getPosition()))
+                .map(l -> new ReferenceLocation(l, false))
                 .collect(Collectors.toList());
 
-        if (foundSymbolInformations.isEmpty()) {
+        // It might be a location on top of a reference instead of a definition.
+        // Ex: Test test; - clicking on Test
+        Set<ReferenceLocation> locations =
+                indexer.getGotoReferenced().keySet().stream()
+                        .filter(l -> paramsUri.toString().equals(l.getUri())
+                                && Ranges.isValid(l.getRange())
+                                && Ranges.contains(l.getRange(), params.getPosition())
+                                && indexer.gotoReferenced(l).isPresent())
+                        .map(l -> new ReferenceLocation(l, true))
+                        .collect(Collectors.toSet());
+        foundReferencedLocations.addAll(locations);
+
+        if (foundReferencedLocations.isEmpty()) {
             return Sets.newHashSet();
         }
 
-        SymbolInformation foundSymbol = foundSymbolInformations.get(0);
-        Set<SymbolInformation> foundReferences = Sets.newHashSet();
+        foundReferencedLocations = foundReferencedLocations.stream()
+                // If there is more than one result, we want the symbol whose range starts the latest, with a secondary
```

Comment:
If two location start at the same position, we want the one who ends first to be selected. It's the most "specific" selection. This is only required because right now some of the symbol/reference locations are not 100% on the name of the var/method/class, so they might overlap.